src/protocol/ipa_prf/boolean_ops/comparison_and_subtraction_low_com.rs

```diff
@@ -0,0 +1,439 @@
+#[cfg(all(test, unit_test))]
+use ipa_macros::Step;
+
+#[cfg(all(test, unit_test))]
+use crate::{
+    error::Error,
+    ff::{ArrayAccess, CustomArray, Expand, Field},
+    protocol::{basics::SecureMul, context::Context, step::BitOpStep, RecordId},
+    secret_sharing::{replicated::semi_honest::AdditiveShare, WeakSharedValue},
+};
+
+#[cfg(all(test, unit_test))]
+#[derive(Step)]
+pub(crate) enum Step {
+    SaturatedSubtraction,
+    MultiplyWithCarry,
+}
+
+///Comparison operation
+/// outputs x>=y
+/// # Errors
+/// propagates errors from multiply
+#[cfg(all(test, unit_test))]
+pub async fn compare_geq<C, XS, YS>(
+    ctx: C,
+    record_id: RecordId,
+    x: &AdditiveShare<XS>,
+    y: &AdditiveShare<YS>,
+) -> Result<AdditiveShare<XS::Element>, Error>
+where
+    C: Context,
+    for<'a> &'a AdditiveShare<XS>: IntoIterator<Item = AdditiveShare<XS::Element>>,
+    YS: WeakSharedValue + CustomArray<Element = XS::Element>,
+    XS: WeakSharedValue + CustomArray + Field,
+    XS::Element: Field + std::ops::Not<Output = XS::Element>,
+{
+    let mut carry = AdditiveShare(XS::Element::ONE, XS::Element::ONE);
+    // we don't care about the subtraction, we just want the carry
+    let _ = subtraction_circuit(ctx, record_id, x, y, &mut carry).await;
+    Ok(carry)
+}
+
+///Comparison operation
+/// outputs x>y
+/// # Errors
+/// propagates errors from multiply
+#[cfg(all(test, unit_test))]
+pub async fn compare_gt<C, XS, YS>(
+    ctx: C,
+    record_id: RecordId,
+    x: &AdditiveShare<XS>,
+    y: &AdditiveShare<YS>,
+) -> Result<AdditiveShare<XS::Element>, Error>
+where
+    C: Context,
+    for<'a> &'a AdditiveShare<XS>: IntoIterator<Item = AdditiveShare<XS::Element>>,
+    YS: WeakSharedValue + CustomArray<Element = XS::Element>,
+    XS: WeakSharedValue + CustomArray + Field,
+    XS::Element: Field + std::ops::Not<Output = XS::Element>,
+{
+    let mut carry = AdditiveShare::<XS::Element>::ZERO;
+    // we don't care about the subtraction, we just want the carry
+    let _ = subtraction_circuit(ctx, record_id, x, y, &mut carry).await;
+    Ok(carry)
+}
+
+///non-saturated unsigned integer subtraction
+/// subtracts y from x, Output has same length as x (carries and indices of y too large for x are ignored)
+/// when y>x, it computes `(x+"XS::MaxValue")-y`
+/// # Errors
+/// propagates errors from multiply
+#[cfg(all(test, unit_test))]
+pub async fn integer_sub<C, XS, YS>(
+    ctx: C,
+    record_id: RecordId,
+    x: &AdditiveShare<XS>,
+    y: &AdditiveShare<YS>,
+) -> Result<AdditiveShare<XS>, Error>
+where
+    C: Context,
+    for<'a> &'a AdditiveShare<XS>: IntoIterator<Item = AdditiveShare<XS::Element>>,
+    YS: WeakSharedValue + CustomArray<Element = XS::Element>,
+    XS: WeakSharedValue + CustomArray + Field,
+    XS::Element: Field + std::ops::Not<Output = XS::Element>,
+{
+    let mut carry = AdditiveShare(XS::Element::ONE, XS::Element::ONE);
```

Comment:
I'm a bit confused by this syntax:
```
AdditiveShare(XS::Element::ONE, XS::Element::ONE);
```
Doesn't that mean the (left, right) are *both* one for all 3 helpers? I guess that's one way to express a secret sharing of 1 in a field of size 2... Is that a correct interpretation of what's going on?

Reply:
Yes! We need to set secret shared value (named carry) to 1.